### ISSUE TITLE
vo=gpu/gpu-next: cache files are generated in config directory after mpv closes

### ISSUE THREAD

**Issue by jiangzhenjerry** [mpv-player/mpv#11898]
### Important Information

- mpv version: [MPV 2023-07-07 13:04](https://github.com/zhongfly/mpv-winbuild/releases/tag/2023-07-07)
- Platform and Version: Windows 11 x64
- Source of the mpv binary: [mpv-x86_64-v3-20230707-git-6ed521f.7z](https://github.com/zhongfly/mpv-winbuild/releases/download/2023-07-07/mpv-x86_64-v3-20230707-git-6ed521f.7z)

### Reproduction steps

1. Start with an empty/default configuration.
2. Start mpv by `mpv --vo=gpu-next` (Note: no need to play any video file here; just start the gui.)
3. Close mpv.

### Expected behavior

No extra file should be generated.

### Actual behavior

A file called `libplacebo.cache` is generated in the configuration folder. In my case, it's generated under `portable_config`.

Is this a bug or a new expected behavior in the latest version? If it's the latter, is there a way to disable it?

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/11985075/output.txt)

### Sample files

N/A as this issue does not need a video file to reproduce. On the other hand, a sample `libplacebo.cache` file is attached below. (I zipped it as GitHub doesn't allow uploading a raw `.cache` file.)

[libplacebo.cache.zip](https://github.com/mpv-player/mpv/files/11985091/libplacebo.cache.zip)

**Comment by Dudemanguy** [comment]
Saving cache is the expected behavior and new default. You can disable it with `--no-gpu-shader-cache` and `--no-icc-cache`.

**Comment by haasn** [comment]
I think the real issue here is not that it saves a cache, but that the cache is saved to the *portable_config* directory.

Shouldn't we save it to some windows equivalent of /tmp or ~/.cache?

**Comment by Dudemanguy** [comment]
>Shouldn't we save it to some windows equivalent of /tmp or ~/.cache?

Sure, someone would just need to add it in `path-win.c`. Whatever the right directory is.

**Comment by Hrxn** [comment]
@jiangzhenjerry are you using `portable_config`?

**Comment by hooke007** [comment]
```
[   0.002][d][global] home path: '' -> 'E:/temp/mpv/portable_config'
[   0.002][d][global] user path: '~~home/' -> 'E:/temp/mpv/portable_config'
```

**Comment by jiangzhenjerry** [comment]
Thanks everyone for the hints.

@Hrxn Yes I am. If I don't use `portable_config` then the cache file is by default saved in `%AppData%\mpv`.

**Comment by Hrxn** [comment]
> I think the real issue here is not that it saves a cache, but that the cache is saved to the _portable_config_ directory.
> 
> Shouldn't we save it to some windows equivalent of /tmp or ~/.cache?

Well, then the placement of `libplacebo.cache` seems to be correct, no?

**Comment by jiangzhenjerry** [comment]
> Well, then the placement of `libplacebo.cache` seems to be correct, no?

I think **haasn** was suggesting to place the cache in a temp folder, not the configuration folder.

**Comment by Dudemanguy** [comment]
>Well, then the placement of libplacebo.cache seems to be correct, no?

mpv has an internal concept of cache directories now but I only implemented it in linux. For everything else, this is going to fall back to the configuration folder (which is the old behavior btw). [QT](https://doc.qt.io/qt-6/qstandardpaths.html) suggests `"C:/Users/<USER>/AppData/Local/<APPNAME>/cache"` for cacheLocation on windows although I don't know if that should be different if one is using the portable_config or not.

**Comment by Hrxn** [comment]
Yeah, well, no... placing it in one of the two "official" Windows Temp directories would kinda defeat its purpose, I think?
Because you want to have `libplacebo.cache`, it speeds up operation of mpv, no benefit if this gets deleted all the time by programs/scripts/etc that clean the temp directories..

**Comment by hooke007** [comment]
I think in config dir is reasonable. But a sub-folder is necessary (ref `watch_later`), or the config dir looks mess.

**Comment by philipl** [comment]
`AppData/Local/` seems like the right place. Windows has the concept of the roaming directory (which could be network shared in an Enterprise configuration) and the local directory, which stays on the local machine, and would contain non-critical state that can be lost without ill effect.

**Comment by jiangzhenjerry** [comment]
Just to avoid confusion:

`%AppData%` = `C:\Users\<USER>\AppData\Roaming`
`%LocalAppData%` = `C:\Users\<USER>\AppData\Local`

> `AppData/Local/` seems like the right place.

Based on the concept you brought up, currently the cache is **not** saved in the right place, because it is saved in `%AppData%\mpv`, that is, `C:\Users\<USER>\AppData\Roaming\mpv`

**Comment by Jules-A** [comment]
> `AppData/Local/` seems like the right place. 

That would just make it annoying if you wanted to manually clear it since all the other stuff like watch later data is stored in roaming... I think it really just needs to be in a sub-folder, so `%AppData%\mpv\cache\`

**Comment by Dudemanguy** [comment]
Just FYI, watch later corresponds internally to `state` now (another linux thing, `~/.local/state/mpv`). That can quite easily be moved somewhere else if it makes sense. It literally just takes someone modifying `path-win.c` so it returns something different for the `cache` and `state` types.

**Comment by philipl** [comment]
Well, watch later should probably go into `Local` as well...

**Comment by haasn** [comment]
> > `AppData/Local/` seems like the right place.
> 
> That would just make it annoying if you wanted to manually clear it since all the other stuff like watch later data is stored in roaming... I think it really just needs to be in a sub-folder, so `%AppData%\mpv\cache\` If you are going to change it to local then watch later should probably also move there.

Why would you want to manually clear it? And this argument also applies to Linux, where `~/.cache` is a separate dir from `~/.config` and `~/.local/share`. I don't see what's "annoying" about it.

Actually, I would find it much, much more annoying if mpv stored the cache in the config directory. What if my config directory is a git repo?!

**Comment by Jules-A** [comment]
> Why would you want to manually clear it? And this argument also applies to Linux, where `~/.cache` is a separate dir from `~/.config` and `~/.local/share`. I don't see what's "annoying" about it.
> 

Well, I guess there's no reason really, except maybe after experimenting with different shaders and modifications, just clearing to get rid of redundant cache. That's just a me issue though.... Watch later data is a more common case for clearing as it doesn't just store playback state but stuff like audio-delay. 

> Actually, I would find it much, much more annoying if mpv stored the cache in the config directory. What if my config directory is a git repo?!

If it was in a separate folder then just git ignore the folder?

**Comment by rossy** [comment]
Yeah, I agree that if caches and watch_later go in `~/.local` in Linux, they should go in `%LOCALAPPDATA%` in Windows, and `%APPDATA%` should just be for config that can "roam" between computers. No reason to make mpv behave differently between Windows and Linux.

I don't think it's clear what we should do about `portable_config` though. I think people expect [portable][1] [apps][2] not to leave files behind on the host computer, so it's not unreasonable to say that the caches could go in `portable_config` when it's being used.

[1]: https://en.wikipedia.org/wiki/Portable_application#Portable_Windows_applications
[2]: https://portableapps.com/about/what_is_a_portable_app#guidelines

**Comment by xM8WVqaG** [comment]
> > `AppData/Local/` seems like the right place.
> 
> That would just make it annoying if you wanted to manually clear it since all the other stuff like watch later data is stored in roaming... I think it really just needs to be in a sub-folder, so `%AppData%\mpv\cache\` If you are going to change it to local then watch later should probably also move there.

In the context of "when to roam" (which is the difference between these directories on Windows), it's quite likely that when you actually roam, different computers will have different hardware and thus have different caches. However, in the case of _watch later_ it's likely that you would want to persist the progress of files between different computers.

In both cases, it's very useful to be able to safely delete a directory knowing it only contains automatically generated files, and know you're not about to nuke actual config.

**Comment by Hrxn** [comment]
> Yeah, I agree that if caches and watch_later go in `~/.local` in Linux, they should go in `%LOCALAPPDATA%` in Windows, and `%APPDATA%` should just be for config that can "roam" between computers. No reason to make mpv behave differently between Windows and Linux.
> 
> I don't think it's clear what we should do about `portable_config` though. I think people expect [portable](https://en.wikipedia.org/wiki/Portable_application#Portable_Windows_applications) [apps](https://portableapps.com/about/what_is_a_portable_app#guidelines) not to leave files behind on the host computer, so it's not unreasonable to say that the caches could go in `portable_config` when it's being used.

Yeah, I'm definitely in favor of this as well..

- `%APPDATA%` : Configuration stuff, everything that can be moved, synced, etc.
- `%LOCALAPPDATA%` : Automatically generated data, caches, everything that ismachine specific and therefore not supposed to be synced
- `portable_config`: To be truly portable, everything is self-contained, i.e. it should be a subdirectory at that specific location

**Comment by Zabooby** [comment]
Shader cache files are generating in my config too, however I'm using vo=gpu and instead of a singular cache file being generated there are multiple see #11917, can anyone reproduce this behaviour?

PS: I also support the idea that these files should be in its own folder.

**Comment by Dudemanguy** [comment]
That's normal. vo=gpu has multiple cache files instead of one.

**Comment by Zabooby** [comment]
Is there any tangible benefit of having shader cache on by default if, (I'm guessing) most people don't use loads of complicated shaders in the first place?

**Comment by kasper93** [comment]
Yes, there is benifit even with default configuration. And there is no real downside, except current fiasco with target directory, but this is a bug, which shouldn't affect decision whether to enable cache or not.

**Comment by hooke007** [comment]
Maybe off topic, if restoring cache in RAM, should it be faster than read them in disk?(espeacially for those ‘medieval’ devices)

**Comment by avih** [comment]
> Just to be clear, this won't actually use cache

Sure. My bad. Let me rephrase.

Can someone please list some common use cases (not the rare ones please) where the cache is created and used, and then estimate very roughly how much time is saved in ms, and when is this time saving happening? (only on startup? also in other cases?)

**Comment by Jules-A** [comment]
> Can someone please list some common use cases (not the rare ones please) where the cache is created and used, and then estimate very roughly how much time is saved in ms, and when is this time saving happening? (only on startup? also in other cases?)

I know you wanted to know what it's like without extra shaders but with my Anime4k preset I'm seeing 3.1s without and 0.35s with to launch and play a 1080p h264 file.
I'll test without shaders in a sec.

**Comment by hooke007** [comment]
> Times were just measured with stopwatch so not very accurate.

For `startup`, you could run mpv with --log-file and then check it.
```
#cahce 1st time
[   2.834][d][cplayer] starting video playback
#cahce 2nd time
[   0.225][d][cplayer] starting video playback
```

**Comment by Jules-A** [comment]
> > Times were just measured with stopwatch so not very accurate.
> 
> For `startup`, you could run mpv with --log-file and then check it.

Ooh, didn't think of that....
```
Shaders ect. enabled
[   3.690][d][cplayer] starting video playback
[   0.756][d][cplayer] starting video playback (cache)

none
[   0.651][v][cplayer] starting audio playback
[   0.304][v][cplayer] starting audio playback (cache)
```

Though I had some other programs open this time so slightly higher.

**Comment by avih** [comment]
So sounds like with "heavy" shaders it can save 2-3 seconds, which is very useful.

While with a more default-ish config it can save 200-300 ms, which is OK I guess.

Personally I don't think I ever saw startup times of 500ms (without cache that I know of), but that could depend on many things.

Anyway, thanks for the numbers.

**Comment by erickyun** [comment]
> > Yeah, I agree that if caches and watch_later go in `~/.local` in Linux, they should go in `%LOCALAPPDATA%` in Windows, and `%APPDATA%` should just be for config that can "roam" between computers. No reason to make mpv behave differently between Windows and Linux.
> > I don't think it's clear what we should do about `portable_config` though. I think people expect [portable](https://en.wikipedia.org/wiki/Portable_application#Portable_Windows_applications) [apps](https://portableapps.com/about/what_is_a_portable_app#guidelines) not to leave files behind on the host computer, so it's not unreasonable to say that the caches could go in `portable_config` when it's being used.
> 
> Yeah, I'm definitely in favor of this as well..
> 
> * `%APPDATA%` : Configuration stuff, everything that can be moved, synced, etc.
> * `%LOCALAPPDATA%` : Automatically generated data, caches, everything that is machine specific and therefore not supposed to be synced
> * `portable_config`: To be truly portable, everything is self-contained, i.e. it should be a subdirectory at that specific location

+1
